### ISSUE TITLE
Godot 4.0 support ?

### ISSUE THREAD

**Issue by RemyCampanals** [Lexpartizan/Godot_sky_shader#18]
Hello @Lexpartizan 

First of all, thank you for your time and for your script. It's very good ! Second, i'm french so sorry in advance if my english is not perfect :p

I tried to make it work in Godot 4.0 for two days but couldn't. I don't have any error, but just nothing happen. The CloudlessSky shader and the Clouds shader doesn't seem to do anything for the Sprite2D inside the viewport.

Here what happened when i tried to support the scripts to Godot 4.0 :

**First for the Sky.gd :** 
1. The `extends Spatial` now is `extends Node3D`, it seems simple but seems to have consequences since there's methods which are not accessible anymore.
2. The new setget syntax has been reworked. Do know why but i had to change how you change the time of day setup (and how you change the `hours`, `minutes` and `seconds` variables) because you have a infinite loop. Doesn't seems to bother Godot 3.4 but 4.0 crash instantaneously.
3. The `lerp` method in 3.4 can lerp from a Color to another Color, not anymore. It's from a float to a float. So i found on internet how to create a `lerp_color`, not sure if it works perfectly (i'm sure it doesn't).
4. The node `Viewport` doesn't exist anymore, now it's `SubViewport`. It seems to work differently ? Maybe that's why my viewport preview seems empty and the whole sky system doesn't work
5. The Sun_Moon object which were just a `DirectionalLight` is now `DirectionalLight3D` and doesn't seems to extends the Spatial node. So when we change the time of day, it seems the methods `look_at_from_position` does nothing. No rotation of the sun...

**The Result :**
![image](https://user-images.githubusercontent.com/31853292/184120794-0ecf5de6-8f84-49ce-9b2b-65ea2fbd7808.png)

**Here the GIST :**
https://gist.github.com/RemyCampanals/044fe42ed5c336780d21bc77ca67c0ac

Thanks to have taken the time to read my issue and hope we can find a solution :)
If you prefer to work on a project in the Godot 4.0 environment rather than just the file, i maybe can prepare that for you :)

**Comment by Rinocrosser** [comment]
Thank you for all the reviewed issues,
Sadly this one isn't supported, it was THE BEST SKY FOR GODOT 3.

**Comment by Lexpartizan** [comment]
Please forgive me for not supporting the adaptation of this sky for Godot 4. My country has unleashed a big war in Europe and I am not able to focus on anything right now.

**Comment by RemyCampanals** [comment]
> Please forgive me for not supporting the adaptation of this sky for Godot 4. My country has unleashed a big war in Europe and I am not able to focus on anything right now. I just can't work on Godot 4 right now because I'm going crazy.

Oh you’re from Russian ?
Don’t worry i understand. 

For now, the team for Godot 4.0 are making a lot of changes for shaders, so it’s best for everyone to wait until it’s « stable ». And for you to rest :)
Take care

**Comment by Lexpartizan** [comment]
Thanks. I'll try to work on the shader a bit later. Let's hope that the math can't be broken and the main problems will be in the script that needs to be adapted. And at least I wrote the script myself and there is no complicated mathematics there.

**Comment by RemyCampanals** [comment]
When i have time, i’ll upload on github a version « working a little » on Godot 4.0 so you don’t have to scratch your head on the depreacated methods ans the new ways to do things on Godot 4.0.

But i’m afraid the problem lies on the math used in the shader and how the UV is wrapped in the panorama sky, but i can be wrong (as i’m still learning shaders)

**Comment by Lexpartizan** [comment]
Thanks. I'll try to work on the shader a bit later. Let's hope that the math can't be broken and the main problems will be in the script that needs to be adapted. And at least I wrote the script myself and there is no complicated mathematics there.

Unfortunately, the rest is not so simple, I have not done anything for a year because of what is happening. I quit my job in order not to bring income to the country. One person with unlimited power and madness can bring death to hundreds of thousands and even, in the future, millions of people. And the surrounding people are either afraid to oppose, as I am afraid, or are delighted with what is happening (the absolute majority). I hope the war will end with the punishment of criminals.

**Comment by Lexpartizan** [comment]
> Thanks. I'll try to work on the shader a bit later. Let's hope that the math can't be broken and the main problems will be in the script that needs to be adapted. And at least I wrote the script myself and there is no complicated mathematics there.
> 
> Unfortunately, the rest is not so simple, I have not done anything for a year because of what is happening. I quit my job in order not to bring income to the country. One person with unlimited power and madness can bring death to hundreds of thousands and even, in the future, millions of people. And the surrounding people are either afraid to oppose, as I am afraid, or are delighted with what is happening (the absolute majority). I hope the war will end with the punishment of criminals.

Yes, we need to see. Sky shaders are now a separate case and use some kind of their own transformation for projection. So, maybe now you can not calculate the projection or somehow you need to avoid it.

**Comment by RemyCampanals** [comment]
Ok, so here we go :

- An empty project with my attempt to "make it work" :
[godot_sky_shader_4_0](https://github.com/RemyCampanals/godot_sky_shader_4_0)
- Links to download the version of Godot 4.0 beta 14 :
https://downloads.tuxfamily.org/godotengine/4.0/beta14/

The things you need to know :
- All the setget from the gd script have to be redone entirely. 
Instead of a simple 
`export (int) var clouds_resolution: int=1024 setget set_clouds_resolution`
we need to write the setget like this
```
@export var clouds_resolution: int = 1024:
	get:
		return clouds_resolution;
	set(value):
		clouds_resolution = value;
		set_clouds_resolution(clouds_resolution);
```
- I deleted the callbacks from the `set_hours`, `set_minutes` and `set_seconds` and i keep the one from `time_of_day_setup`. It creates a infinite loop otherwise. In godot 3.4 it's not problem but it is in Godot 4.0 :(
- Playing with the clouds_resolution and the sky_resolution seems to break the shader for unknow reason. So i let them at 1024 and 2048 respectively.
- Only clouds2d fonction in the shaders "displays clouds" (when the settings of the quality is between 1 and 19. The clouds3d show nothing, but maybe it tries to but below the horizon.

Here's what my result on Godot 4.0 beta 14 :
![Sky Shader](http://image.noelshack.com/fichiers/2023/04/2/1674575617-capture-d-ecran-20230124-163833.jpg)

**Comment by RemyCampanals** [comment]
> > Thanks. I'll try to work on the shader a bit later. Let's hope that the math can't be broken and the main problems will be in the script that needs to be adapted. And at least I wrote the script myself and there is no complicated mathematics there.
> > Unfortunately, the rest is not so simple, I have not done anything for a year because of what is happening. I quit my job in order not to bring income to the country. One person with unlimited power and madness can bring death to hundreds of thousands and even, in the future, millions of people. And the surrounding people are either afraid to oppose, as I am afraid, or are delighted with what is happening (the absolute majority). I hope the war will end with the punishment of criminals.
> 
> Yes, we need to see. Sky shaders are now a separate case and use some kind of their own transformation for projection. So, maybe now you can not calculate the projection or somehow need to avoid it.

Yes, and i think in the future we need to use the new Sky Shader. But for now i still not try to understand how it works. Maybe the projection is calculated automaticaly.

Once more, take your time and take care of your self. There is no emergency for making it work.
Thanks in advance for the time you took for me :)

**Comment by Lexpartizan** [comment]
> > > Thanks. I'll try to work on the shader a bit later. Let's hope that the math can't be broken and the main problems will be in the script that needs to be adapted. And at least I wrote the script myself and there is no complicated mathematics there.
> > > Unfortunately, the rest is not so simple, I have not done anything for a year because of what is happening. I quit my job in order not to bring income to the country. One person with unlimited power and madness can bring death to hundreds of thousands and even, in the future, millions of people. And the surrounding people are either afraid to oppose, as I am afraid, or are delighted with what is happening (the absolute majority). I hope the war will end with the punishment of criminals.
> > 
> > 
> > Yes, we need to see. Sky shaders are now a separate case and use some kind of their own transformation for projection. So, maybe now you can not calculate the projection or somehow need to avoid it.
> 
> Yes, and i think in the future we need to use the new Sky Shader. But for now i still not try to understand how it works. Maybe the projection is calculated automaticaly.
> 
> Once again, take your time and take care of yourself. There is no emergency for making this shader works with Godot 4.0. Thanks in advance for your time you already provide me :)

![image](https://user-images.githubusercontent.com/48552343/222697415-cd98d30c-0084-4ca3-8a4a-a3d9141132cd.png)

I finally forced myself to sit down for this shader. And I already used special SKY shaders. So there is some progress. I think everything will take 2-3 weeks, because I still need to rewrite the code to the script and I constantly encounter the fact that the solutions I am used to do not work in GODOT 4.

**Comment by Lexpartizan** [comment]
> Thank you for all the reviewed issues, Sadly this one isn't supported, it was THE BEST SKY FOR GODOT 3.
For GODOT 4
https://github.com/Lexpartizan/Godot_4_sky_shader

**Comment by RemyCampanals** [comment]
Wow already ?
You rock ! 👌

So excited to go home in order to test this ! 
Thank you in advance 🙌
And take care !

**Comment by RemyCampanals** [comment]
First of all, well done and thank you !

BUT :p
![image](https://user-images.githubusercontent.com/31853292/224169341-5e6b95d1-5a12-4612-a5c5-d054d8e0e939.png)
I have no clouds in day or night and there is some glitches below the horizon.

Since you provide me a screen where everything works, maybe you didn't push the right version on github ? 
(Do you prefer i create an issue in your SkyShader for Godot 4.0 github instead of here ?)

**Comment by Lexpartizan** [comment]
> 

![image](https://user-images.githubusercontent.com/48552343/224266933-b0d64963-5f89-48ad-988a-cb878869443a.png)

check textures applaying in shader

**Comment by RemyCampanals** [comment]
I just opened your project, it seems to have all the parameters set :)

But i don't understand, are you using one single image file in order to "display the moon", "the thunder" and "the noise for clouds" ?
If it's the case, i think it's better to separate. Single Responsibility Principle 😇

**Comment by Lexpartizan** [comment]
> Я только что открыл ваш проект, вроде все параметры выставлены :)
> 
> Но я не понимаю, вы используете один файл изображения, чтобы «отобразить луну», «гром» и «шум для облаков»? Если это так, я думаю, что лучше расстаться. Принцип единой ответственности😇

Give a screenshot of your scene and specify the shader parameters, please.

I just downloaded this project and opened it separately, everything is fine. Try re-importing the textures as I used the Basis Universalis settings for them.

**Comment by Lexpartizan** [comment]
> I just opened your project, it seems to have all the parameters set :)
> 
> But i don't understand, are you using one single image file in order to "display the moon", "the thunder" and "the noise for clouds" ? If it's the case, i think it's better to separate. Single Responsibility Principle 😇

Give a screenshot of your scene "sky" and specify the shader parameters, please.

I just downloaded this project and opened it separately, everything is fine. Try re-importing the textures as I used the Basis Universalis settings for them.

**Single Responsibility Principle**
one shader - one texture. Less memory used.

**Comment by Lexpartizan** [comment]
> I just opened your project, it seems to have all the parameters set :)
> 
> But i don't understand, are you using one single image file in order to "display the moon", "the thunder" and "the noise for clouds" ? If it's the case, i think it's better to separate. Single Responsibility Principle 😇

You have discord? Lexicon#0978

**Comment by RemyCampanals** [comment]
> > I just opened your project, it seems to have all the parameters set :)
> > But i don't understand, are you using one single image file in order to "display the moon", "the thunder" and "the noise for clouds" ? If it's the case, i think it's better to separate. Single Responsibility Principle 😇
> 
> You have discord? Lexicon#0978

Yes i have ! I'm EmryX#7031, i added you :)

**Comment by umutbalkan** [comment]
I also downloaded the godot 4 version and opened it, it seems like clouds are beneath the horizon with some problems. Textures and parameters have set
![image](https://user-images.githubusercontent.com/8284138/233783083-29dc5c32-fb46-4f3d-aed6-1f45a8a13d18.png)

**Comment by RemyCampanals** [comment]
> I also downloaded the godot 4 version and opened it, it seems like clouds are beneath the horizon with some problems. Textures and parameters have set
> ![image](https://user-images.githubusercontent.com/8284138/233783083-29dc5c32-fb46-4f3d-aed6-1f45a8a13d18.png)

Are you on Nvidia card (RTX 20, 30 or 40 series) ?

**Comment by Lexpartizan** [comment]
> > Я также скачал версию godot 4 и открыл ее, кажется, что облака над горизонтом с некоторыми проблемами. Текстуры и параметры установлены
> > ![изображение](https://user-images.githubusercontent.com/8284138/233783083-29dc5c32-fb46-4f3d-aed6-1f45a8a13d18.png)
> 
> Используете ли вы видеокарту Nvidia (RTX 20, 30 или 40 серии)?
My card is 1050. Yes, with the new nvidia graphics cards there are problems with this. I created an issue.
https://github.com/godotengine/godot/issues/76343

**Comment by Lexpartizan** [comment]
@RemyCampanals 
@umutbalkan 

My card is 1050. Yes, with the new nvidia graphics cards there are problems with this. I created an issue.
https://github.com/godotengine/godot/issues/76343
There is no solution yet.
